src/utils.c

```diff
@@ -110,30 +117,231 @@ SEXP colnamesInt(SEXP x, SEXP cols, SEXP check_dups) {
+  int protecti=0;
+  if (!isString(mode) || length(mode)!=1 || STRING_ELT(mode, 0)==NA_STRING)
+    error(_("internal error: mode argument has to be character, non-NA of length 1")); // # nocov
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=NULL
+  if (!isNewList(x))
+    error(_("'x' argument must be data.table"));
+  bool mode_j=false, mode_sd=false;
+  if (strcmp(CHAR(STRING_ELT(mode, 0)), "j")==0)
+    mode_j = true;
+  else if (strcmp(CHAR(STRING_ELT(mode, 0)), ".SDcols")==0)
+    mode_sd = true;
+  else
+    error(_("Internal error: invalid 'mode' argument in exprCols function, should have been caught before. please report to data.table issue tracker.")); // # nocov
+  // TODO in colnamesInt: handle inverse selection: !cols_var, !"V2", !V2, !paste0("V",2:3)
+  // but not !V3:V2, we could handle that as well but do we really want?
+  // we already have tests for !(V3:V2) so we do want to support that
+  // TODO test for !!V3
+  bool inverse = isLanguage(expr) && (CAR(expr)==sym_bang || CAR(expr)==sym_minus); // length(expr)==2, otherwise V3-V2 will pick it up
+  if (inverse)
+    expr = CADR(expr);
+  SEXP sym_parenthesis = install("(");
+  while (isLanguage(expr) && CAR(expr)==sym_parenthesis)
+    expr = CADR(expr);
+  SEXP sym_brackets = install("{");
+  while (isLanguage(expr) && CAR(expr)==sym_brackets && length(expr)==2)
+    expr = CADR(expr); // verify #376, 2121.2
+  SEXP sym_eval = install("eval");
+  if (isLanguage(expr) && CAR(expr)==sym_eval) {
+    
+  }
+  //SEXP attribute_hidden do_allnames(SEXP call, SEXP op, SEXP args, SEXP env)
+  if (isNull(expr))
+    error(_("columns selection is NULL")); // expr=((NULL))
+  // non-evaluated case: V3:V2
+  // single symbol V2 may or may not be evaluated, see below
+  if (isLanguage(expr) && CAR(expr)==sym_colon) {  // 3:2, V3:V2, min(V3):min(V2)
+    SEXP lhs = CADR(expr), rhs = CADDR(expr);
+    if (isSymbol(lhs) && isSymbol(rhs)) { // V3:V2
+      if (inverse)
+        error("column range selection by col1:col2 is not supported with combination of negation via '!' or '-'");
+      lhs = colnamesInt(x, ScalarString(PRINTNAME(lhs)), ScalarLogical(false), ScalarLogical(false)); // may raise error if lhs column does not exists
+      rhs = colnamesInt(x, ScalarString(PRINTNAME(rhs)), ScalarLogical(false), ScalarLogical(false));
+      if (!isInteger(lhs) || !isInteger(rhs) || length(lhs)!=1 || length(rhs)!=1 || LOGICAL(lhs)[0]==NA_LOGICAL || LOGICAL(rhs)[0]==NA_LOGICAL)
+        error(_("internal error: LHS and RHS of `:` call should be integer non-NA scalars already")); // # nocov
+      UNPROTECT(protecti);
+      return range_int(INTEGER(lhs)[0], INTEGER(rhs)[0]);
+    } else {
+      // evaluates later on: 3:2, f(V3):f(V2)
+      //Rf_PrintValue(lhs);
+      //Rf_PrintValue(rhs);
+    }
+  }
+  // patterns will not evaluate as well because we wrap expr into quote(expr) before calling eval on do_patterns call
```

Comment:
We may be able to get away from ```do_patterns```. There is a ```cols``` argument that potentially subsets the names of the DT but it isn't in any of the tests and I am not sure of the use case. In other words, it may be good to deprecate the ```cols``` argument and only have ```patterns(...)``` instead of ```patterns(..., cols = character(0)```. Deprecating should not affect many users because ```patterns``` is not exported, and again, it's not in the testing units.

If we get away from ```do_patterns``` and/or change the args, we could drop straight to ```grep``` in the C level with ```do_grep```. This is what it looks like in R:

```
## from 
cols = Reduce(intersect, do_patterns(colsub, x_names))

## to 
pat_sub = as.list(colsub)[-1L]
cols = Reduce(intersect, lapply(lapply(pat_sub, eval, parent.frame(2L)), grep, x_names))
```
https://github.com/SurajGupta/r-source/blob/a28e609e72ed7c47f6ddfbb86c85279a0750f0b7/src/main/grep.c#L742

Reply:
`patterns` function is exported, at least as a manual entry. It is also well documented, including `cols` argument. Unfortunately there are no unit tests for that functionality, which I already marked in tests.Rraw file, here in this PR. Note there are examples that uses this feature.

Reply:
It is documented but ```data.table::patterns``` does not work. It seems to only work in two places, ```melt``` and ```.SDcols```. In the context of the only two places it's evaluated, only the names of the data.table provided makes sense. ```melt(x, patterns("^x"), c("random","character", "vector")``` would not return intended results, nor would ```melt(as.data.table(iris), patterns("Sepal"), c("Species", "Sepal.1", "Sepal.2")```

However, I did not mean to do an off-the-cuff deprecation. My implementation in R is around 30-50 microseconds. Yours is typically around 5x faster. Does it make sense to further complicate things to make ```patterns()``` searches faster than microseconds? Probably not.

Reply:
in any case it will be best to have a separate issue for that if needed.